pkg/k8s/k8s.go

```diff
@@ -475,44 +476,57 @@ func CmdDelK8s(ctx context.Context, c calicoclient.Interface, epIDs utils.WEPIde
-		// Delete the WorkloadEndpoint object from the datastore, passing revision information from the
-		// queried resource above in order to prevent conflicts.
-		switch err := err.(type) {
-		case cerrors.ErrorResourceDoesNotExist:
-			// Log and proceed with the clean up if WEP doesn't exist.
-			logger.WithField("endpoint", wep).Info("Endpoint object does not exist, no need to clean up.")
-		case cerrors.ErrorResourceUpdateConflict:
-			// This case means the WEP object was modified between the time we did the Get and now,
-			// so it's not a safe Compare-and-Delete operation, so log and abort with the error.
-			// Returning an error here is with the assumption that k8s (kubelet) retries deleting again.
-			logger.WithField("endpoint", wep).Warning("Error deleting endpoint: endpoint was modified before it could be deleted.")
-			return fmt.Errorf("error deleting endpoint: endpoint was modified before it could be deleted: %v", err)
-		case cerrors.ErrorOperationNotSupported:
-			// KDD does not support WorkloadEndpoint deletion, the WEP is backed by the Pod and the
-			// deletion will be handled by Kubernetes. This error can be ignored.
-			logger.WithField("endpoint", wep).Info("Endpoint deletion will be handled by Kubernetes deletion of the Pod.")
-		default:
-			return err
+			// The WorkloadEndpoint doesn't exist for some reason. We should still try to clean up any IPAM allocations
+			// if they exist, so continue DEL processing.
+			logger.WithField("WorkloadEndpoint", epIDs.WEPName).Warning("WorkloadEndpoint does not exist in the datastore, moving forward with the clean up")
+		} else if wep.Spec.ContainerID != "" && args.ContainerID != wep.Spec.ContainerID {
+			// If the ContainerID is populated and doesn't match the CNI_CONTAINERID provided for this execution, then
+			// we shouldn't delete the workload endpoint. We identify workload endpoints based on pod name and namespace, which means
+			// we can receive DEL commands for an old sandbox for a currently running pod. However, we key IPAM allocations based on the
+			// CNI_CONTAINERID, so we should still do that below for this case.
+			logger.WithField("WorkloadEndpoint", wep).Warning("CNI_CONTAINERID does not match WorkloadEndpoint ConainerID, don't delete WEP.")
+		} else if _, err = c.WorkloadEndpoints().Delete(
+			ctx,
+			wep.Namespace,
+			wep.Name,
+			options.DeleteOptions{
+				ResourceVersion: wep.ResourceVersion,
+				UID:             &wep.UID,
+			},
+		); err != nil {
+			// Delete the WorkloadEndpoint object from the datastore, passing revision information from the
+			// queried resource above in order to prevent conflicts.
+			switch err := err.(type) {
+			case cerrors.ErrorResourceDoesNotExist:
+				// Log and proceed with the clean up if WEP doesn't exist.
+				logger.WithField("endpoint", wep).Info("Endpoint object does not exist, no need to clean up.")
+			case cerrors.ErrorResourceUpdateConflict:
```

Comment:
If we hit a UID conflict here do we actually want to retry?

e.g.,

- We do a GET
- pod is deleted / recreated with new UID
- We get conflict
- We do a GET (now retrieving the new UID)
- We properly delete the WEP this time (removing any IP addresses, which might actually be valid now?)

Honestly, seems _really_ unlikely since to be a problem it would require a CNI ADD occurring in between the retries. But in theory possible?

Reply:
I guess in KDD mode we're protected by the deletiontimestamp and the fact that we don't rely on the annotations (always fall back to the status fields), and in etcd mode this isn't an issue. So probably fine.

Reply:
ErrorResourceUpdateConflict means the revision was wrong; not the UID.  If the UID is wrong then we get ErrorResourceDoesNotExist